File: src/09/day9.py
from functools import reduce

def get_input(filename: str) -> list[int]:
    with open(filename, 'rt') as f:
        for line in f:
            yield [int(x) for x in line.strip().split()]

def get_layers(numbers: list[int], index: int) -> list[int]:
    layers = []
    while any(x != 0 for x in numbers):
        layers.append(numbers[index])
        numbers = [numbers[i+1] - numbers[i] for i in range(len(numbers)-1)]
    return layers

def part1(filename: str) -> int:
    return sum(sum(get_layers(numbers, -1)) for numbers in get_input(filename))

def part2(filename: str) -> int:
    diff = lambda itr: reduce(lambda acc, x: x - acc, itr, 0)
    return sum(diff(reversed(get_layers(numbers, 0))) for numbers in get_input(filename))

if __name__ == '__main__':
    print(part1('day9_input.txt'))
    print(part2('day9_input.txt'))

File: src/09/day9.rs
fn get_input(filename: &str) -> Vec<Vec<i64>> {
    let lines = std::fs::read_to_string(filename).expect("File not found");
    let mut values: Vec<Vec<i64>> = Vec::new();
    for line in lines
        .trim()
        .split('\n')
        .map(|x|
            x.split_whitespace()
            .map(|v| v.parse::<i64>().unwrap())
            .collect::<Vec<_>>()
        ) {
        values.push(line);
    }
    return values;
}

fn get_layers(numbers: &Vec<i64>, part_two: bool) -> Vec<i64> {
    let mut layers: Vec<i64> = Vec::new();
    let mut differences: Vec<i64> = numbers.to_vec();
    while differences.iter().any(|&x| x != 0) {
        layers.push(if part_two { differences[0] } else { differences[differences.len()-1] });
        differences = (0..(differences.len()-1))
            .map(|i| differences[i+1] - differences[i])
            .collect::<Vec<_>>();
    }
    return layers;
}

fn part1(filename: &str) -> i64 {
    return get_input(filename)
        .iter()
        .map(|n| get_layers(n, false).iter().sum::<i64>())
        .sum::<i64>();
}

fn part2(filename: &str) -> i64 {
    return get_input(filename)
        .iter()
        .map(|n| get_layers(n, true).into_iter().rev().reduce(|acc, x| x - acc).unwrap())
        .reduce(|acc, x| acc + x).unwrap();
}

fn main() {
    println!("{}", part1("day9_input.txt"));
    println!("{}", part2("day9_input.txt"));
}